torchtune/datasets/_packed.py

```diff
@@ -0,0 +1,78 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Dict, List, Optional, Tuple
+
+from torch.utils.data import Dataset
+from tqdm import tqdm
+
+
+class PackedDataset(Dataset):
+    """
+    Performs greedy sample packing on a provided dataset. This is done as a single
+    preprocessing step before training begins. Shuffling is done outside of this
+    class on packed samples as part of the dataloader.
+
+    The class loads, tokenizes, and packs examples on initialization - no tokenization is done during training.
+
+    The general flow on initialization is: load tokenized sample -> add to buffer ->
+        when buffer is long enough, add to self.samples.
+
+    During training, returns self.samples[idx] as input and label.
+
+    Args:
+        ds (Dataset): dataset to sample pack. This should return a tuple of tokenized
+            inputs and labels.
+        max_seq_len (int): Maximum number of tokens to pack
+        max_rows (Optional[int]): maximum number of samples to pack. Default is None, which will pack as many samples as possible.
+    """
+
+    def __init__(
+        self,
+        ds: Dataset,
+        max_seq_len: int,
+        max_rows: Optional[int] = None,
+    ) -> None:
```

Comment:
Ideally, if this is an `IterableDataset` it will be much more efficient (e.g., when doing pre-training - dataset is too large to fit in memory for packing). But i guess that can be pretty challenging to do, especially considering shuffling.

Reply:
Yeah the IterableDataset would be a more efficient approach, but the design of this in a distributed setting with shuffling and packing support with modular, extensible design that can generalize to all our datasets is challenging. This is currently in the works (cc @SLR722) but may take some time so I wanted to add packing for simple map style datasets that can load in memory for now.

Reply:
I agree! This will a long-term thing IMO - i think load in memory is pretty sufficient more 99% SFT workload :)